File: server/src/handlers/sync.py
"""
Sync endpoint for OpenMemory Server (v0.3.1b).

Provides delta sync for agent reconnection.
Returns all facts for a user since a given sequence_id.

Spec: docs/specs/openmemory/server.md v0.3.1b section 4, 8.2
"""
import logging
from fastapi import APIRouter, Depends, HTTPException, status, Header, Query
from pydantic import BaseModel, Field
from typing import Optional, List

from ..auth import hash_auth_key
from ..db import get_db, Database
from ..dependencies import get_current_user

logger = logging.getLogger(__name__)

router = APIRouter(tags=["sync"])


# ============ Response Models ============

class SyncedFactJSON(BaseModel):
    """A fact returned by the sync endpoint."""
    id: str
    sequence_id: Optional[int] = None
    encrypted_blob: str  # hex-encoded
    blind_indices: List[str]
    decay_score: float
    is_active: bool
    version: int
    source: str
    content_fp: Optional[str] = None
    agent_id: Optional[str] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class SyncResponseJSON(BaseModel):
    """Response from GET /sync."""
    success: bool
    facts: List[SyncedFactJSON] = []
    latest_sequence: int = 0
    has_more: bool = False
    error_code: Optional[str] = None
    error_message: Optional[str] = None


# ============ Endpoint ============

@router.get("/sync", response_model=SyncResponseJSON)
async def sync(
    since_sequence: int = Query(0, description="Return facts with sequence_id > this value"),
    limit: int = Query(1000, ge=1, le=10000, description="Max facts to return"),
    user_id: str = Depends(get_current_user),
    db: Database = Depends(get_db)
):
    """
    Delta sync endpoint for agent reconnection (v0.3.1b).

    Returns all facts for the authenticated user where sequence_id > since_sequence.
    Used by agents after coming online to pull changes made by other agents.

    Response includes:
    - facts: array of facts with all metadata
    - latest_sequence: current highest sequence_id for this user
    - has_more: true if more facts beyond limit (client should paginate)
    """
    try:
        facts, latest_seq, has_more = await db.get_facts_since_sequence(
            user_id=user_id,
            since_sequence=since_sequence,
            limit=limit
        )

        synced_facts = []
        for fact in facts:
            synced_facts.append(SyncedFactJSON(
                id=fact.id,
                sequence_id=getattr(fact, 'sequence_id', None),
                encrypted_blob=fact.encrypted_blob.hex(),
                blind_indices=fact.blind_indices,
                decay_score=fact.decay_score,
                is_active=fact.is_active,
                version=fact.version,
                source=fact.source,
                content_fp=getattr(fact, 'content_fp', None),
                agent_id=getattr(fact, 'agent_id', None),
                created_at=fact.created_at.isoformat() if fact.created_at else None,
                updated_at=fact.updated_at.isoformat() if fact.updated_at else None,
            ))

        return SyncResponseJSON(
            success=True,
            facts=synced_facts,
            latest_sequence=latest_seq,
            has_more=has_more,
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Sync failed: {e}", exc_info=True)
        return SyncResponseJSON(
            success=False,
            error_code="INTERNAL_ERROR",
            error_message="Sync failed"
        )

File: client/src/api/sync.ts
/**
 * Client Reconnection Protocol (v0.3.1b)
 *
 * Provides delta sync against the /sync endpoint and
 * local fact reconciliation for agent reconnection.
 *
 * Spec: docs/specs/totalreclaw/server.md v0.3.1b section 8.2
 *
 * Protocol:
 * 1. Agent comes online
 * 2. GET /sync?since_sequence={last_known_sequence}
 * 3. Build set of server fingerprints: { content_fp -> fact_id }
 * 4. For each local pending fact:
 *    a. If content_fp in server set -> skip
 *    b. Else -> POST /store (server also checks, but pre-filtering avoids round trips)
 * 5. Update local last_known_sequence
 */

// ============================================================================
// Types
// ============================================================================

/**
 * A fact returned by the /sync endpoint.
 */
export interface SyncedFact {
  id: string;
  sequence_id: number | null;
  encrypted_blob: string;
  blind_indices: string[];
  decay_score: number;
  is_active: boolean;
  version: number;
  source: string;
  content_fp: string | null;
  agent_id: string | null;
  created_at?: string;
  updated_at?: string;
}

/**
 * A fact pending local push to the server.
 */
export interface LocalPendingFact {
  id: string;
  content_fp: string | undefined;
  plaintext: string;
  // Other fields needed for store request omitted for simplicity;
  // the caller is responsible for assembling the full store payload.
}

/**
 * Result of sync.
 */
export interface SyncResult {
  facts: SyncedFact[];
  latestSequence: number;
  hasMore: boolean;
}

/**
 * Result of local reconciliation.
 */
export interface ReconciliationResult {
  /** Local facts that match server (should be skipped) */
  skip: LocalPendingFact[];
  /** Local facts that are genuinely new (should be pushed) */
  push: LocalPendingFact[];
}

/**
 * Config for the sync client.
 */
export interface SyncClientConfig {
  serverUrl: string;
  /** Optional: inject a fetch implementation for testing */
  fetchImpl?: typeof fetch;
}

// ============================================================================
// SyncState
// ============================================================================

/**
 * Tracks the sync watermark for an agent.
 */
export class SyncState {
  public lastSequence: number;
  public lastSyncAt: Date | null;

  constructor(lastSequence: number = 0) {
    this.lastSequence = lastSequence;
    this.lastSyncAt = null;
  }

  update(latestSequence: number): void {
    this.lastSequence = latestSequence;
    this.lastSyncAt = new Date();
  }

  toJSON(): string {
    return JSON.stringify({
      lastSequence: this.lastSequence,
      lastSyncAt: this.lastSyncAt?.toISOString() ?? null,
    });
  }

  static fromJSON(json: string): SyncState {
    const data = JSON.parse(json);
    const state = new SyncState(data.lastSequence ?? 0);
    state.lastSyncAt = data.lastSyncAt ? new Date(data.lastSyncAt) : null;
    return state;
  }
}

// ============================================================================
// SyncClient
// ============================================================================

/**
 * HTTP client for the /sync endpoint.
 */
export class SyncClient {
  private serverUrl: string;
  private fetchImpl: typeof fetch;

  constructor(config: SyncClientConfig) {
    this.serverUrl = config.serverUrl.replace(/\/$/, '');
    this.fetchImpl = config.fetchImpl ?? fetch;
  }

  /**
   * Fetch one page of facts since the given sequence.
   */
  async syncSince(
    sinceSequence: number,
    authKey: string,
    limit: number = 1000
  ): Promise<SyncResult> {
    const url = `${this.serverUrl}/v1/sync?since_sequence=${sinceSequence}&limit=${limit}`;
    const response = await this.fetchImpl(url, {
      method: 'GET',
      headers: {
        Authorization: `Bearer ${authKey}`,
        Accept: 'application/json',
      },
    });

    if (!response.ok) {
      throw new Error(`Sync failed: HTTP ${response.status}`);
    }

    const data = await response.json() as {
      success: boolean;
      facts?: SyncedFact[];
      latest_sequence?: number;
      has_more?: boolean;
      error_message?: string;
    };
    if (!data.success) {
      throw new Error(`Sync failed: ${data.error_message}`);
    }

    return {
      facts: data.facts ?? [],
      latestSequence: data.latest_sequence ?? 0,
      hasMore: data.has_more ?? false,
    };
  }

  /**
   * Fetch ALL facts since the given sequence, auto-paginating.
   */
  async syncAllSince(
    sinceSequence: number,
    authKey: string,
    limit: number = 1000
  ): Promise<SyncResult> {
    const allFacts: SyncedFact[] = [];
    let currentSeq = sinceSequence;
    let latestSequence = 0;

    // eslint-disable-next-line no-constant-condition
    while (true) {
      const page = await this.syncSince(currentSeq, authKey, limit);
      allFacts.push(...page.facts);
      latestSequence = page.latestSequence;

      if (!page.hasMore || page.facts.length === 0) {
        break;
      }

      // Advance the cursor to the last sequence in this page
      const lastFact = page.facts[page.facts.length - 1];
      if (lastFact.sequence_id != null) {
        currentSeq = lastFact.sequence_id;
      } else {
        break; // Cannot paginate without sequence_id
      }
    }

    return {
      facts: allFacts,
      latestSequence,
      hasMore: false,
    };
  }
}

// ============================================================================
// Reconciliation
// ============================================================================

/**
 * Reconcile local pending facts against server state.
 *
 * For each local fact:
 * - If content_fp matches a server fact -> skip (already stored)
 * - Otherwise -> push (genuinely new)
 *
 * @param serverFacts - Facts from the /sync response
 * @param localFacts - Local pending facts to reconcile
 * @returns Which local facts to skip and which to push
 */
export function reconcileLocalFacts(
  serverFacts: SyncedFact[],
  localFacts: LocalPendingFact[]
): ReconciliationResult {
  // Build server fingerprint set
  const serverFpSet = new Set<string>();
  for (const fact of serverFacts) {
    if (fact.content_fp) {
      serverFpSet.add(fact.content_fp);
    }
  }

  const skip: LocalPendingFact[] = [];
  const push: LocalPendingFact[] = [];

  for (const local of localFacts) {
    if (local.content_fp && serverFpSet.has(local.content_fp)) {
      skip.push(local);
    } else {
      push.push(local);
    }
  }

  return { skip, push };
}
